rspirv/sr/types.rs

```diff
@@ -16,262 +16,54 @@ use crate::spirv;
 
 use super::{Constant, Decoration, Token};
 
-/// The class to represent a SPIR-V type.
-///
-/// Derived types hold `TypeToken`s of their base types instead of direct
-/// references or `Rc` references, because of the difficulty or impossibility
-/// of labelling explicit lifetimes or back references.
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Type {
-    pub(in crate::sr) ty: TypeEnum,
-    /// Sets of decorations.
-    pub(in crate::sr) decorations: Vec<Decoration>,
+#[derive(Clone, Debug)]
+pub struct StructMember {
+    pub token: Token<Type>,
+    pub decorations: Vec<Decoration>,
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(in crate::sr) struct StructMember {
-    pub(in crate::sr) token: Token<Type>,
-    pub(in crate::sr) decorations: Vec<Decoration>,
-}
+include!("autogen_types.rs");
 
-impl StructMember {
-    pub(in crate::sr) fn new(token: Token<Type>) -> Self {
-        StructMember {
-            token,
-            decorations: Vec::new(),
+impl PartialEq for Type {
```

Comment:
The spec is really dodgy here, Section 2.8, page 22 states:
>Two different type <id>s form, by definition, two different types. It is valid to declare multiple aggregate type <id>s
having the same opcode and operands. This is to allow multiple instances of aggregate types with the same structure to be decorated differently. (Different decorations are not required; two different aggregate type <id>s are allowed to have identical declarations and decorations, and will still be two different types.) Non-aggregate types are different: It is invalid to declare multiple type <id>s for the same scalar, vector, or matrix type. That is, non-aggregate type declarations must all have different opcodes or operands. (Note that non-aggregate types cannot be decorated in ways that affect their type.)

This makes no mention of opaque types, I assume they also are never equal due to their opaqueness but I'm not sure.